text/0000-read-all.md

```diff
@@ -0,0 +1,50 @@
+- Feature Name: read_all
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's Write trait has write_all, which attempts to write an entire
+buffer.  This proposal adds read_all, which attempts to read a fixed
+number of bytes into a given buffer.
+
+# Motivation
+
+The new read_all method will allow programs to read from disk without
+having to write their own read loops.  Most Rust programs which need
+to read from disk will prefer this to the plain read function.  Many C
+programs have the same need, and solve it the same way (e.g. git has
+read_in_full).  Here's one example of a Rust library doing this:
+https://github.com/BurntSushi/byteorder/blob/master/src/new.rs#L184
+
+# Detailed design
+
+The read_all function will take a mutable, borrowed slice of u8 to
+read into, and will attempt to fill that entire slice with data.
+
+It will loop, calling read() once per iteration and attempting to read
+the remaining amount of data.  If read returns EINTR, the loop will
+retry.  If there are no more bytes to read (as signalled by a return
+of Ok(0) from read()), a new error type, ErrorKind::ReadZero, will be
+returned.  In the event of another error, that error will be
+returned. After a read call returns having successfully read some
+bytes, the total number of bytes read will be updated.  If that
+total is equal to the size of the buffer, read will return
+successfully.
```

Comment:
1. Consistency with write_all.
2. io::Result<uint> is either Ok(uint) or Error.  read_all has only one kind of ok result: "yep, we read the whole thing".  If instead we returned Ok(number of bytes read), then every caller would have to check that the number of bytes read matched their expectations.  In other words, we are optimizing for the common case of "either this file has the data, or we're doomed".  And because the error on EOF is distinct from other errors, callers who care can choose to match on it and handle it separately.  But I expect that most callers will not care.